fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py

```diff
@@ -751,24 +755,95 @@ def set_field(tokenizers_info_cfg: List, name: str, key: str, val: Any) -> List:
+
+        # and special tokens
+
+        new_tokenize_special_tokens = self.get_subtokenizer_added_tokens(t_json)
+        
+        # update the ids 
+        updated_vocab, next_index = ModularTokenizer.remap_vocab(
+            vocab=old_vocab,
+            starting_index=next_index,
+        )
 
-        # 1. Load the tokenizer
-        # 2. Create a list of its special tokens
-        # 3. Call self.add_special_tokens with the new list
-        # 4. Add a new entry to self.tokenizers_info
-        # 5. Call remap_vocab on the new tokenizer json, with the updated special tokens of self (similar to lines 131..142)
+        # Take the special tokens as they appear in the first tokenizer inside the modular tokenizer
+        first_tok_key = list(self.tokenizers_info.keys())[0]
+        all_special_token_structs = self.tokenizers_info[first_tok_key]["json_instance"]["added_tokens"]
 
-        # self.build_inner_decoder()
-        # if self._max_possible_token_id is not None:
-        #     if self._get_max_mapped_id() > self._max_possible_token_id:
-        #         raise Exception(
-        #             f"tokenizer remapping resulted in IDs greater (max_id={self._get_max_mapped_id()}) than max_possible_id ({self._max_possible_token_id}). Reinitialize the modular tokenizer with larger max_possible_id"
-        #         )
+        # check if we are really adding any tokens from the new tokenizer
+        new_special_tokens = set(new_tokenize_special_tokens) - set([tok["content"] for tok in all_special_token_structs])
+         # in the begging of the list, with their id's
+
+        t_json["model"]["vocab"] = {sp["content"]:sp["id"] for sp in all_special_token_structs}
+        # and finally we add the updated vocabulary.
+
+        t_json["model"]["vocab"].update(updated_vocab)
+
+        # we also need to add the tokens to the added_tokens section as special tokens
+        t_json["added_tokens"]=copy.deepcopy(all_special_token_structs)
+
+        # create the tokenizer instance TODO: do we need this?
+        json_str = json.dumps(t_json)
+        tokenizer_inst = tokenizers.Tokenizer.from_str(json_str)
+        new_tokenizer_info["tokenizer_inst"]=tokenizer_inst
+
+        self.tokenizers_info[new_tokenizer_info["name"]] = new_tokenizer_info
+        
+        added_tokens = get_additional_tokens(subset=["special", "task"])
+        if new_special_tokens: # did we add them in the special_tokens.py file - which is the correct way to work for extended tokenizers
+            new_special_tokens = set(new_special_tokens) - set(added_tokens)
+        if new_special_tokens:
+            warn(
+                    f"Added tokenizer {new_tokenizer_info['name']} came with {len(new_special_tokens)} new special tokens: {','.join(new_special_tokens)}"
+                )
+            added_tokens += new_special_tokens
+
+        # we update the special tokens but do not save here.  remember to save yourself.
+        self.update_special_tokens(
```

Comment:
I think it's best to put self.update_special_tokens after self.build_inner_decoder. It should be fine as is, but self.update_special_tokens assumes that the modular tokenizer is consistent and usable. This may not be the case before build_inner_decoder is called

Reply:
Done